text/vm-2.0.md

```diff
@@ -0,0 +1,228 @@
+
+What’s more, the use of `mut` now corresponds to whether the JavaScript VM’s state can be modified, which is a much more natural use of mutability annotations. This also makes it easier to explain *why* the `vm` argument should be annotated as `mut` for passing mutable references to other functions.
+
+Finally, this design sets the stage for [a more complete design of the macro syntax](https://gist.github.com/dherman/d1d6d62e4e44ffbee1e1eebd51b53f06#file-classes_two_point_oh-rs-L48-L63), which should be fleshed out in a subsequent "Classes 2.0" RFC. In particular, the new syntax should allow an optional identifier (conventionally, this would normally be named `vm`), and an optional `mut` annotation, before the formal parameters list. Taking a representation of the state of the JavaScript VM, which is required for passing to many Neon APIs, again provides a higher level of abstraction than a representation of a function activation in the form of a `FunctionCall` struct. This makes for a gentler slope from the highest-level convenience syntaxes to the more expressive forms.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The `Vm` trait
+
+Most Neon functions take an extra parameter in the form of a mutable reference to an implementation of the `Vm` trait. This parameter roughly represents access to the JavaScript virtual machine. For example, the `Object::get` method for getting a property of a JavaScript object requires an extra `Vm` parameter:
+
+```rust
+let foo = obj.get(&mut vm, "foo");
+```
+
+A good way to think about the `Vm` object is that it represents **the power to manipulate the JavaScript virtual machine**. In particular, mutable access to the `Vm` makes it possible to modify objects, execute JavaScript code, and perform garbage collection.
+
+A powerful implication of this is that an API that does *not* take a mutable reference to a `Vm` **can never modify the JavaScript virtual machine**! This allows Neon to provide the ability to safely *lock* the virtual machine temporarily, allowing Rust code to inspect and modify the internals of native objects or typed arrays.
+
+## Locking the VM
+
+Some type of JavaScript values—such as typed arrays, or instances of Neon classes—implement an `Inspect` trait, which allows Rust to access their internals. In order to get access to these internals, you first need to *take out a lock* on the JavaScript VM. The `Vm::lock` method “freezes” the VM and produces a temporary `Lock` object, which you can then use to inspect the object.
+
+For example, a simple Neon function that takes a single `ArrayBuffer` object and sets its first byte to 0 would be implemented by calling `vm.lock()` and then `lock.inspect()` with the resulting lock object:
+
+```rust
+fn zero_first_byte(mut vm: impl Vm) -> JsResult<Handle<JsNumber>> {
```

Comment:
Is there a reason that this needs to rely on unstable features? Could this just be `fn zero_first_byte<T: Vm>(mut vm: T)`?

Reply:
Oh I was just using the shorthand because it makes me happy :P There's no inherent reliance on it. I should eliminate that to avoid confusion.